middleware/dnstap/msg/msg.go

```diff
@@ -24,8 +45,34 @@ type Data struct {
 	TimeSec     uint64
 }
 
-func (d *Data) FromRequest(r request.Request) error {
-	switch addr := r.W.RemoteAddr().(type) {
+// HostPort decodes into Data any string returned by dnsutil.ParseHostPortOrFile.
+func (d *Data) HostPort(addr string) (err error) {
+	ip, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return
+	}
+	p, err := strconv.ParseUint(port, 10, 32)
+	if err != nil {
+		return
+	}
+	d.Port = uint32(p)
+
+	if ip := net.ParseIP(ip); ip != nil {
```

Comment:
reverse this:
```
ip := net.ParseIP(ip)
if ip == nil {
   return fmt.Errorf("not an in: %s", ip)
}
```
and then the test

Reply:
It's a trick so I can use the same variable name for a different type under the `if` scope.